os/fs/procfs/fs_procfspminfo.c

```diff
@@ -0,0 +1,224 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <pm.h>
+#include <pm_metrics.h>
+#include <string.h>
+#include <fcntl.h>
+#include <assert.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <tinyara/kmalloc.h>
+#include <tinyara/fs/fs.h>
+#include <tinyara/fs/procfs.h>
+#include <tinyara/pm/pm.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define PMINFO_LINELEN 256
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct pminfo_file_s {
+	struct procfs_file_s base;	/* Base open file structure */
+	unsigned int linesize;		/* Number of valid characters in line[] */
+	char line[PMINFO_LINELEN];	/* Pre-allocated buffer for formatted lines */
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+/* File system methods */
+
+static int pminfo_open(FAR struct file *filep, FAR const char *relpath, int oflags, mode_t mode);
+static int pminfo_close(FAR struct file *filep);
+static ssize_t pminfo_read(FAR struct file *filep, FAR char *buffer, size_t buflen);
+
+/****************************************************************************
+ * Private Variables
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Variables
+ ****************************************************************************/
+
+/* See fs_mount.c -- this structure is explicitly externed there.
+ * We use the old-fashioned kind of initializers so that this will compile
+ * with any compiler.
+ */
+
+const struct procfs_operations pminfo_operations = {
+	pminfo_open,				/* open */
+	pminfo_close,				/* close */
+	pminfo_read,				/* read */
+	NULL,						/* write */
+	NULL,						/* dup */
```

Comment:
This is one of file operation, also working with fd, so dup & stat are necessary.